addons/l10n_se_ocr/models/account_move.py

```diff
@@ -0,0 +1,67 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import api, fields, models, _
+from odoo.exceptions import UserError
+
+class AccountMove(models.Model):
+    _inherit = 'account.move'
+
+    def _get_invoice_reference_se_ocr2(self, reference):
+        self.ensure_one()
+        return self.env['l10n_se.luhn'].append(reference)
+
+    def _get_invoice_reference_se_ocr3(self, reference):
+        self.ensure_one()
+        return self.env['l10n_se.luhn'].append(reference + str(len(reference)+2)[:1])
+
+    def _get_invoice_reference_se_ocr4(self, reference):
+        self.ensure_one()
+
+        ocr_length = self.journal_id.l10n_se_invoice_ocr_length
+
+        if ocr_length < 5:
+            raise UserError(_("OCR Reference Number length need to be greater than 5. Please correct settings under invoice journal settings."))
+
+        if len(reference)+1 > ocr_length:
+            raise UserError(_("OCR Reference Number length is greater than allowed. Allowed length in invoice journal setting is %s.") % str(ocr_length))
+
+        return self.env['l10n_se.luhn'].append(reference.ljust(ocr_length-1, '0'))
+
+    def _get_invoice_reference_se_ocr2_invoice(self):
+        self.ensure_one()
+        return self._get_invoice_reference_se_ocr2(str(self.id))
+
+    def _get_invoice_reference_se_ocr3_invoice(self):
+        self.ensure_one()
+        return self._get_invoice_reference_se_ocr3(str(self.id))
+
+    def _get_invoice_reference_se_ocr4_invoice(self):
+        self.ensure_one()
+        return self._get_invoice_reference_se_ocr4(str(self.id))
+
+    def _get_invoice_reference_se_ocr2_partner(self):
+        self.ensure_one()
+        return self._get_invoice_reference_se_ocr2(self.partner_id.ref if str(self.partner_id.ref).isdecimal() else str(self.partner_id.id))
+
+    def _get_invoice_reference_se_ocr3_partner(self):
+        self.ensure_one()
+        return self._get_invoice_reference_se_ocr3(self.partner_id.ref if str(self.partner_id.ref).isdecimal() else str(self.partner_id.id))
+
+    def _get_invoice_reference_se_ocr4_partner(self):
+        self.ensure_one()
+        return self._get_invoice_reference_se_ocr4(self.partner_id.ref if str(self.partner_id.ref).isdecimal() else str(self.partner_id.id))
+
+    @api.onchange('partner_id')
+    def _onchange_partner_id(self):
+        """ If Vendor Bill and Vendor OCR is set, add it. """
+        if self.partner_id and self.type == 'in_invoice' and self.partner_id.l10n_se_check_vendor_ocr and self.partner_id.l10n_se_default_vendor_ocr:
```

Comment:
We should remove self.partner_id.l10n_se_check_vendor_ocr from this if. It will be much clearer for the user what it should do.

Reply:
Sure, however I would recommend to clear the default value if the check box is triggered on the partner. If this isn't cleared it will put the value on the invoice and the user will not really understand why. It could easily be done with a onchange for the l10n_se_check_vendor_ocr field in res_partner. What do you think?

Reply:
or I could just remove the invisible attribute in view.

Change the name to l10n_se_default_payment_reference and add a check_ocr on the luhn check in res_partner. 

That way the default_payment_reference could be used for other things like messages.

Reply:
@daniel-stenlov I still want to know why we need l10n_se_check_vendor_ocr here.  If the default value is set we can put it.

Reply:
@jco-odoo, this has been removed the code looks like this from account_move.py

```
    @api.onchange('partner_id')
    def _onchange_partner_id(self):
        """ If Vendor Bill and Vendor OCR is set, add it. """
        if self.partner_id and self.type == 'in_invoice' and self.partner_id.l10n_se_default_vendor_payment_ref:
            self.invoice_payment_ref = self.partner_id.l10n_se_default_vendor_payment_ref
        return super(AccountMove, self)._onchange_partner_id()
```